### PR TITLE
[Model] Add HrmTextForCausalLM (Hierarchical Reasoning Model — Text)

### PR DESCRIPTION
**Pull request by abcd1927**
HRM-Text shipped in [transformers 5.9.0](https://pypi.org/project/transformers/5.9.0/) (released 2026-05-20, merged to `main` via [huggingface/transformers#46025](https://github.com/huggingface/transformers/pull/46025)). The model performs a hierarchical recurrent forward over two transformer stacks (`H` slow, `L` fast) inside nested H/L cycle loops; each recurrence step consumes a distinct KV cache slot, matching transformers' `cycle_offset` formula. Attention uses a sigmoid gate (Qwen3Next-style) and a fused gate+q+k+v projection on disk.

This PR is fully scoped to the new model file, the registry, and the supported-models doc — no changes to attention backends, the AttentionType enum, or Attention itself. Pipeline parallelism and LoRA are deferred to follow-up PRs (each currently raises at construction).

## Architecture

- `HrmTextAttention` keeps `gate_proj` (`ColumnParallelLinear`) separate from `qkv_proj` (`QKVParallelLinear`) so both shard cleanly along the head axis under TP. The HF on-disk schema fuses them into a single `attn.gqkv_proj.weight` (rows concatenated as `[gate | q | k | v]`); `HrmTextForCausalLM.load_weights` splits that tensor at load time before passing it to `AutoWeightsLoader`. Each recurrence step gets its own `Attention` instance under a `nn.ModuleDict`, with a unique `prefix` (`...layers.{global_idx}.attn`), so vLLM allocates a distinct KV cache slot per cycle step. Total slots match `H_cycles * (L_cycles + 1) * num_layers_per_stack`, equal to `config.num_hidden_layers` after HF's `__post_init__` inflation.

- `_create_hrm_attention_backend` wraps an attention backend's metadata builder so `causal=False` is set unconditionally on every build. Mirrors `EncoderOnlyAttention`'s `subclass_attention_backend` pattern, but keeps the KV cache (the recurrent forward needs to reuse it).

- `HrmTextModel.forward` matches HF main exactly: nested H-outer L-inner loop, `z_L = L(z_L + z_H)` then `z_H = H(z_H + z_L)`. The Python loop is unrolled by `@support_torch_compile` since cycle counts are config constants.

## Why unconditional causal=False

vLLM v1 schedules continuous batching. Even without chunked prefill, a single `build()` call can see a *mixed* batch where some requests are decoding and others are entering prefill on this step. Gating on `is_prefilling.all()` keeps `causal=True` in that mixed case, silently running the newly-prefilling requests as pure causal — which diverges from HRM-Text's PrefixLM training distribution.

Unconditional `causal=False` is correct because:
- Prefill rows (query_len=N): bidirectional, matching HF main with `token_type_ids=ones_like(input_ids)`.
- Decode rows (query_len=1): `causal=True` and `causal=False` are mathematically identical (no future tokens to mask).
- FlashAttention varlen kernels apply `causal` per sub-sequence via `cu_seqlens_q`, so a single global flag does not cause cross-prompt contamination.

## Empirical validation

`sapientinc/HRM-Text-1B`, vLLM matches HF transformers main with `token_type_ids=ones_like(input_ids)` to within statistical noise.

| Benchmark | HF (token_type_ids=1) | vLLM (this PR) |
|---|---|---|
| GSM8K, first 50 | 82.0% (41/50) | 82.0% (41/50) |
| GSM8K, full (1319) | — | 84.99% |
| GSM8K, tp=1 vs tp=2 | — | identical (both 82.0% on first 50) |
| DROP, first 100 EM | 85.0% | 85.0%, 100/100 normalized text match |
| DROP, full (9536) | — | 79.98% EM |

## On-disk weight schema

HF transformers writes HrmText checkpoints with these keys (per the "hrm_text" entry in `transformers/conversion_mapping.py`):

```
model.embed_tokens.weight
model.z_L_init
model.{H,L}_module.layers.{i}.attn.gqkv_proj.weight  ([gate|q|k|v] dim 0)
model.{H,L}_module.layers.{i}.attn.o_proj.weight
model.{H,L}_module.layers.{i}.mlp.{gate_up,down}_proj.weight
```

`load_weights` splits `attn.gqkv_proj.weight` into `attn.gate_proj.weight` (first `num_heads * head_dim` rows) and `attn.qkv_proj.weight` (remaining `[q | k | v]` rows) at load time, so vLLM's standard `QKVParallelLinear` weight loader handles TP partitioning of q/k/v without unfusing the gate.

### CODE REVIEW

File: tests/models/registry.py [hunk 1]
```diff
@@ -328,6 +328,12 @@ def check_available_online(
         "hpcai-tech/grok-1", trust_remote_code=True
     ),
     "Grok1ForCausalLM": _HfExamplesInfo("xai-org/grok-2", trust_remote_code=True),
+    "HrmTextForCausalLM": _HfExamplesInfo(
+        "sapientinc/HRM-Text-1B",
+        # HrmText is in transformers main; first release that ships it
+        # has not yet been pinned. Update once the version is known.
+        min_transformers_version="5.9",
```
[the remark this answers is not in the document]
Comment (abcd1927) [reply tests/models/registry.py]: Thanks for the heads-up. A clarification on context:

- transformers latest stable on PyPI is **5.8.1** (released 2026-05-13), not 4.43.x — the bot's training data appears to be out of date here.
- The HrmText model class was only merged to transformers `main` yesterday (2026-05-18, huggingface/transformers#46025) and is **not in any released version yet**. Pinning to `4.43.0` would point at a pre-HrmText release where the import would fail.
- `5.9.0` is a best-effort guess at the next minor release that will ship HrmText. transformers `main` currently has `__version__ = "5.8.0.dev0"` (the dev marker hasn't been bumped past 5.8.1 yet), so the next minor cut should be 5.9.0. I'll update this once it actually ships — flagged in an inline comment in the latest commit.

Updated to `5.9.0` (3-segment, matching the convention used by neighboring entries like `ExaoneMoEForCausalLM`'s `5.1.0`) in f90fd72c906ecd1b761985aa993554f58b57378d.

[the remark this answers is not in the document]
Comment (abcd1927) [reply tests/models/registry.py]: Update: transformers 5.9.0 was released 2026-05-20 and ships HrmText, so `min_transformers_version="5.9.0"` is now confirmed (no longer a guess). Simplified the inline comment in 12aeab035.

File: tests/models/language/generation/test_hrm_text.py
Comment (hmellor) [inline tests/models/language/generation/test_hrm_text.py:1]: I don't think this is necessary, very few models have dedicated test files like this.

File: vllm/model_executor/models/hrm_text.py [hunk 2]
```diff
@@ -0,0 +1,627 @@
...
+        ) -> AttentionMetadata:
+            new_md = copy(common_attn_metadata)
+            new_md.causal = False
+            return super().build(common_prefix_len, new_md, fast_build)
+
+    return subclass_attention_backend(
+        name_prefix="HrmAttention_",
+        attention_backend_cls=underlying_attn_backend,
+        builder_cls=HrmAttentionBuilder,
+    )
```
Comment (hmellor) [inline vllm/model_executor/models/hrm_text.py]: Instead of duplicating this code here, let's update `EncoderOnlyAttention` so it uses `attn_type`

```diff
--- a/vllm/model_executor/layers/attention/encoder_only_attention.py
+++ b/vllm/model_executor/layers/attention/encoder_only_attention.py
@@ -69,17 +69,23 @@ class EncoderOnlyAttention(Attention):
         else:
             kv_cache_dtype = "auto"
 
+        if attn_type is None:
+            attn_type = AttentionType.ENCODER_ONLY
+
         underlying_attn_backend = get_attn_backend(
             head_size,
             dtype,
             kv_cache_dtype,
-            attn_type=AttentionType.ENCODER_ONLY,
+            attn_type=attn_type,
         )
 
         attn_backend = create_encoder_only_attention_backend(underlying_attn_backend)
 
-        if attn_type is not None:
-            assert attn_type == AttentionType.ENCODER_ONLY, (
+        if (
+            attn_type != AttentionType.ENCODER_ONLY
+            and kwargs.pop("raise_on_invalid_attn_type", True)
+        ):
+            raise ValueError(
                 "EncoderOnlyAttention only supports AttentionType.ENCODER_ONLY"
             )
 
@@ -89,7 +95,7 @@ class EncoderOnlyAttention(Attention):
             scale=scale,
             cache_config=cache_config,
             attn_backend=attn_backend,
-            attn_type=AttentionType.ENCODER_ONLY,
+            attn_type=attn_type,
             **kwargs,
         )
```

then in your modelling code you can use:

```py
            self.attn_per_step[str(step)] = EncoderOnlyAttention(
                num_heads=self.num_heads,
                head_size=self.head_dim,
                scale=self.scaling,
                num_kv_heads=self.num_kv_heads,
                cache_config=cache_config,
                quant_config=quant_config,
                attn_type=AttentionType.DECODER,
                prefix=f"{unique_prefix}.attn",
                rais_on_invalid_attn_type=False,
            )
```

File: vllm/model_executor/models/hrm_text.py [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+            quant_config=quant_config,
+            prefix=f"{prefix}.o_proj",
+        )
+
+        rope_parameters = getattr(config, "rope_parameters", None)
+        if rope_parameters is None:
+            rope_parameters = {
+                "rope_type": "default",
+                "rope_theta": getattr(config, "rope_theta", 10000.0),
+            }
```
Comment (hmellor) [inline vllm/model_executor/models/hrm_text.py]: It looks like it's never valid for `rope_parameters` to be `None`. Let's just pass `config.rope_parameters` directly?

File: vllm/model_executor/models/hrm_text.py [hunk 4]
```diff
@@ -0,0 +1,627 @@
...
+        kv_cache_dtype = (
+            cache_config.cache_dtype if cache_config is not None else "auto"
+        )
+        underlying = get_attn_backend(
+            self.head_dim,
+            dtype,
+            kv_cache_dtype,
+            attn_type=AttentionType.DECODER,
+        )
+        wrapped_backend = _create_hrm_attention_backend(underlying)
```
Comment (hmellor) [inline vllm/model_executor/models/hrm_text.py]: With the above changes this block can be deleted

File: vllm/model_executor/models/hrm_text.py [hunk 5]
```diff
@@ -0,0 +1,627 @@
...
+        ``step * num_layers_per_stack + layer_idx_in_stack``, matching
+        the HF transformers ``cycle_offset`` formula in
+        ``modeling_hrm_text.py``.
+    """
+
+    def __init__(
+        self,
+        config: PretrainedConfig,
+        layer_idx_in_stack: int,
+        stack_kind: str,  # "L" or "H"
```
Comment (hmellor) [inline vllm/model_executor/models/hrm_text.py]:
```suggestion
        stack_kind: Literal["L", "H"],
```

File: vllm/model_executor/models/hrm_text.py [hunk 6]
```diff
@@ -0,0 +1,627 @@
...
+        quant_config: QuantizationConfig | None = None,
+        prefix: str = "",
+    ) -> None:
+        super().__init__()
+        self.hidden_size = config.hidden_size
+        # Named `attn` (not `self_attn`) to match HF's on-disk weight schema:
+        # transformers/conversion_mapping.py renames `attn.o_proj` -> `self_attn.o_proj`
+        # and splits `attn.gqkv_proj` into 4 separate self_attn projections at
+        # load-into-HF-model time. We keep the on-disk name to make vLLM's
+        # AutoWeightsLoader work without a custom WeightsMapper.
```
Comment (hmellor) [inline vllm/model_executor/models/hrm_text.py]: In this case (any any others) I think I would prefer to use the same name as in HF src and use the `WeightsMapper`. They now support regex patterns so you can simply copy whatever is on the HF side.

Once Transformers v4 support is dropped I could make `AutoWeightsLoader` always check the HF conversion mapping, but we can't do that just yet.

**Review by hmellor — COMMENTED** [review]
General comment, can we be more explicit with variable names? In the HF src the names are much easier to understand `high_cycle_idx` & `low_cycle_idx` for example.

File: tests/models/language/generation/test_hrm_text.py
Comment (abcd1927) [reply tests/models/language/generation/test_hrm_text.py:1]: Done.

File: vllm/model_executor/models/hrm_text.py (hunk 2 above)
Comment (abcd1927) [reply vllm/model_executor/models/hrm_text.py]: Done. One thing beyond your diff: `EncoderOnlyAttention.get_kv_cache_spec` was hardcoded to `return None`, which skips KV alloc when `attn_type=DECODER`. Added a branch on `self.attn_type`; encoder-only path unchanged.

File: vllm/model_executor/models/hrm_text.py (hunk 3 above)
Comment (abcd1927) [reply vllm/model_executor/models/hrm_text.py]: Done.

File: vllm/model_executor/models/hrm_text.py (hunk 4 above)
Comment (abcd1927) [reply vllm/model_executor/models/hrm_text.py]: Done.

File: vllm/model_executor/models/hrm_text.py (hunk 5 above)
Comment (abcd1927) [reply vllm/model_executor/models/hrm_text.py]: Done.

File: vllm/model_executor/models/hrm_text.py (hunk 6 above)
Comment (abcd1927) [reply vllm/model_executor/models/hrm_text.py]: Done. Replaced the two-module `gate_proj` + `qkv_proj` with a single 4-way `MergedColumnParallelLinear gqkv_proj` matching the on-disk row layout; weight loader auto-splits along the output dim, so `_split_disk_gqkv` is gone. Attribute renamed to `self_attn` via `WeightsMapper`.

**Comment by abcd1927** [comment]
Done. Renamed loop vars and hidden states to match HF. Kept `H_cycles`/`L_cycles` (config fields) and `H_module`/`L_module` (module names) since HF uses the same.

File: vllm/model_executor/layers/attention/prefix_lm_attention.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    while keeping ``attn_type=DECODER`` so a KV cache is still allocated.
+
+    Effect by phase:
+    - Prefill: query tokens attend to each other bidirectionally -- this is
+      where the Prefix LM (non-causal) behavior actually takes effect.
+    - Single-token decode: ``causal=False`` is a no-op. The one new query
+      attends to the whole (frozen) KV cache exactly as a causal decode
+      would, and cached tokens cannot attend back to it, so the output is
+      identical to a causal decoder.
+    """
```
Comment (Isotr0py) [inline vllm/model_executor/layers/attention/prefix_lm_attention.py]: Actually, we have multimodal prefix-lm (partial bidi) support already:
https://github.com/vllm-project/vllm/blob/309385a359b1722ed0f2720540421a3fba08a9ac/vllm/v1/worker/gpu_model_runner.py#L2453-L2479

Perhaps you can reuse this code path with some modifications for text-only models. (Setting `doc_ranges` to prefill range)

**Comment by abcd1927** [comment]
Implemented via the KV cache spec channel as discussed. `PrefixLMAttention.get_kv_cache_spec` tags its spec with `non_causal=True`, and the engine core scans the specs it already collects from all workers (before the scheduler is built) and disables chunked prefill + prefix caching when any layer is non-causal. TP-safe — the signal crosses the worker→engine boundary on the pickled spec, applied to the same config object the scheduler reads.

Verified on H100: GSM8K (50 samples) gives identical accuracy at TP=1 and TP=2, with the auto-disable confirmed in the engine-core logs at both sizes.

@Isotr0py would appreciate your review on the spec-channel approach.

File: vllm/model_executor/layers/attention/prefill_prefix_lm_attention.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    while keeping ``attn_type=DECODER`` so a KV cache is still allocated.
+
+    Effect by phase:
+    - Prefill: query tokens attend to each other bidirectionally -- this is
+      where the Prefix LM (non-causal) behavior actually takes effect.
+    - Single-token decode: ``causal=False`` is a no-op. The one new query
+      attends to the whole (frozen) KV cache exactly as a causal decode
+      would, and cached tokens cannot attend back to it, so the output is
+      identical to a causal decoder.
+    """
```
Comment (Isotr0py) [inline vllm/model_executor/layers/attention/prefill_prefix_lm_attention.py:31]: I prefer to rename this class as `PrefillPrefixLMAttention`, because this will only work for models that treat the whole prefill sequence as bidi prefix, while `PrefixLMAttention` will cause confusion between Paligemma/Gemma4's multimodal prefix bidi attention.

**Review by Isotr0py — APPROVED** [review]

File: vllm/model_executor/layers/attention/prefill_prefix_lm_attention.py (hunk 8 above)
Comment (abcd1927) [reply vllm/model_executor/layers/attention/prefill_prefix_lm_attention.py:31]: Done — renamed to `PrefillPrefixLMAttention` (and the file to `prefill_prefix_lm_attention.py` to match). Thanks!
